src/precomputed_quantities.jl

```diff
@@ -319,18 +370,238 @@ function set_precomputed_quantities!(Y, p, t)
         end
     end
 
+    if turbconv_model isa DiagnosticEDMFX
+        FT = eltype(Y)
+        (; ᶜp_ref, ᶜρ_ref, ᶠu³) = p
+        (; q_tot) = p.ᶜspecific
+        (; ᶜρaʲs, ᶠu³ʲs, ᶜuʲs, ᶜKʲs, ᶜh_totʲs, ᶜq_totʲs, ᶜtsʲs, ᶜρʲs) = p
+        ᶜ∇p_perturb³ = p.ᶜtemp_CT3
+        ᶜ∇Φ³ = p.ᶜtemp_CT3_2
+
+        @. ᶜ∇p_perturb³ = CT3(ᶜgradᵥ(ᶠinterp(ᶜp - ᶜp_ref)))
+        @. ᶜ∇p_perturb³ += CT3(gradₕ(ᶜp - ᶜp_ref))
+        @. ᶜ∇Φ³ = CT3(ᶜgradᵥ(ᶠinterp(ᶜΦ)))
+        @. ᶜ∇Φ³ += CT3(gradₕ(ᶜΦ))
+
+        ρaʲu³ʲ_data = p.temp_data_level
+        ρaʲu³ʲ_datau³ʲ_data = p.temp_data_level_2
+        ρaʲu³ʲ_datah_tot = ρaʲu³ʲ_dataq_tot = p.temp_data_level_3
+
+        for j in 1:n
+            ᶜρaʲ = ᶜρaʲs.:($j)
+            ᶠu³ʲ = ᶠu³ʲs.:($j)
+            ᶜKʲ = ᶜKʲs.:($j)
+            ᶜh_totʲ = ᶜh_totʲs.:($j)
+            ᶜtsʲ = ᶜtsʲs.:($j)
+            ᶜρʲ = ᶜρʲs.:($j)
+            ᶜq_totʲ = ᶜq_totʲs.:($j)
+
+            # boundary condition
+            ρaʲ_level = Fields.field_values(Fields.level(ᶜρaʲ, 1))
```

Comment:
Would it be useful to keep the `\^c` notation for the `level` variables that are defined here and below?

Reply:
I thought about it too. I decided not to keep it in the end, as the `level` variables are `DataLayouts` instead of `Field`( because we take `field_values`), and `\^c` notation is for `Field`.